storage/integration_test.go

```diff
@@ -4472,49 +4522,57 @@ func TestIntegration_Reader(t *testing.T) {
 			}
 			contents[obj] = c
 		}
+
 		// Test Reader. Cache control and last-modified are tested separately, as
 		// the JSON and XML APIs return different values for these.
 		for _, obj := range objects {
-			rc, err := b.Object(obj).NewReader(ctx)
-			if err != nil {
-				t.Errorf("Can't create a reader for %v, errored with %v", obj, err)
-				continue
-			}
-			if !rc.checkCRC {
-				t.Errorf("%v: not checking CRC", obj)
-			}
+			t.Run(obj, func(t *testing.T) {
+				// Test both Read and WriteTo.
+				for _, c := range readCases {
+					t.Run(c.desc, func(t *testing.T) {
+						rc, err := b.Object(obj).NewReader(ctx)
+						if err != nil {
+							t.Fatalf("Can't create a reader for %v, errored with %v", obj, err)
+						}
+						if !rc.checkCRC {
+							t.Errorf("%v: not checking CRC", obj)
+						}
 
-			slurp, err := ioutil.ReadAll(rc)
-			if err != nil {
-				t.Errorf("Can't ReadAll object %v, errored with %v", obj, err)
-			}
-			if got, want := slurp, contents[obj]; !bytes.Equal(got, want) {
-				t.Errorf("Contents (%q) = %q; want %q", obj, got, want)
-			}
-			if got, want := rc.Size(), len(contents[obj]); got != int64(want) {
-				t.Errorf("Size (%q) = %d; want %d", obj, got, want)
-			}
-			if got, want := rc.ContentType(), "text/plain"; got != want {
-				t.Errorf("ContentType (%q) = %q; want %q", obj, got, want)
-			}
-			rc.Close()
+						slurp, err := c.readFunc(rc)
+						if err != nil {
+							t.Errorf("Can't read object %v, errored with %v", obj, err)
+						}
+						if got, want := slurp, contents[obj]; !bytes.Equal(got, want) {
+							t.Errorf("Contents (%q) = %q; want %q", obj, got, want)
+						}
+						if got, want := rc.Size(), len(contents[obj]); got != int64(want) {
+							t.Errorf("Size (%q) = %d; want %d", obj, got, want)
+						}
+						if got, want := rc.ContentType(), "text/plain"; got != want {
+							t.Errorf("ContentType (%q) = %q; want %q", obj, got, want)
+						}
+						rc.Close()
 
-			// Check early close.
-			buf := make([]byte, 1)
-			rc, err = b.Object(obj).NewReader(ctx)
-			if err != nil {
-				t.Fatalf("%v: %v", obj, err)
-			}
-			_, err = rc.Read(buf)
-			if err != nil {
-				t.Fatalf("%v: %v", obj, err)
-			}
-			if got, want := buf, contents[obj][:1]; !bytes.Equal(got, want) {
-				t.Errorf("Contents[0] (%q) = %q; want %q", obj, got, want)
-			}
-			if err := rc.Close(); err != nil {
-				t.Errorf("%v Close: %v", obj, err)
-			}
+						// Check early close.
+						buf := make([]byte, 1)
+						rc, err = b.Object(obj).NewReader(ctx)
+						if err != nil {
+							t.Fatalf("%v: %v", obj, err)
+						}
+						_, err = rc.Read(buf)
+						if err != nil {
+							t.Fatalf("%v: %v", obj, err)
+						}
+						if got, want := buf, contents[obj][:1]; !bytes.Equal(got, want) {
+							t.Errorf("Contents[0] (%q) = %q; want %q", obj, got, want)
+						}
+						if err := rc.Close(); err != nil {
+							t.Errorf("%v Close: %v", obj, err)
+						}
```

Comment:
Suggestion: Can this be moved out one level? I don't think we need to test this for both read funcs.

Reply:
I think it's fine to leave it as is; it doesn't shrink the test matrix at all to make this change.